### PR TITLE
refactor: collapse several array iterators into a single reduce

### PR DESCRIPTION
**Pull request by probablyup**
**What kind of change does this PR introduce?**
refactoring

**Did you add tests for your changes?**
using existing tests

**If relevant, link to documentation update:**
n/a

**Summary**
This piece of code was very complicated and hard to read due to all the chaining. It seemed like a good use case for a `reduce` instead of all extra array iterations.

**Does this PR introduce a breaking change?**
no

**Other information**
n/a

### CODE REVIEW

File: lib/SourceMapDevToolPlugin.js [hunk 1]
```diff
@@ -43,57 +74,43 @@ class SourceMapDevToolPlugin {
...
 							}
 						}
-						task.modules = modules;
-						return task;
-					}).forEach(task => {
-						tasks.push(task);
-					});
+
+						return queue;
+					}, tasks);
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: `}, []);`

I guess you can create the array inline.

File: lib/SourceMapDevToolPlugin.js [hunk 2]
```diff
@@ -43,57 +74,43 @@ class SourceMapDevToolPlugin {
...
-						}
-					}).filter(Boolean).map(task => {
-						const modules = task.sourceMap.sources.map(source => {
-							const module = compilation.findModule(source);
-							return module || source;
-						});
-						for(const module of modules) {
-							if(!moduleToSourceNameMapping.get(module)) {
-								moduleToSourceNameMapping.set(module, ModuleFilenameHelpers.createFilename(module, moduleFilenameTemplate, requestShortener));
+					tasks = chunk.files.reduce((queue, file) => {
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: `const tasks = ...`

... and create the variable here.

File: lib/SourceMapDevToolPlugin.js [hunk 3]
```diff
@@ -43,57 +74,43 @@ class SourceMapDevToolPlugin {
...
+
+		const matchObject = ModuleFilenameHelpers.matchObject.bind(undefined, options);
+
 		compiler.plugin("compilation", compilation => {
 			new SourceMapDevToolModuleOptionsPlugin(options).apply(compilation);
+
 			compilation.plugin("after-optimize-chunk-assets", function(chunks) {
 				const moduleToSourceNameMapping = new Map();
-				const tasks = [];
+				let tasks = [];
```
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: ... to omit this.

File: lib/SourceMapDevToolPlugin.js (hunk 2 above)
Comment (probablyup) [inline lib/SourceMapDevToolPlugin.js]: This seems like a different behavior from the original code? If we create the tasks array inline with const inside the chunk forEach, that doesn't concatenate the tasks from each chunk onto the master queue anymore for processing.

Unless I misunderstood. There weren't unit tests here so I was trying to be careful.

**Comment by sokra** [comment]
Thanks

File: lib/SourceMapDevToolPlugin.js (hunk 2 above)
Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: Yep, I was wrong. Sorry

Comment (sokra) [inline lib/SourceMapDevToolPlugin.js]: btw. this part of the PR is covered by tests. Only the caching thing above is not covered. 

btw2. I can recommend the Codecov Extension which highlights covered lines...

**Pull request merged**